packages/loopring_v3/contracts/amm/libamm/AmmExitRequest.sol

```diff
@@ -0,0 +1,212 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "../../lib/EIP712.sol";
+import "../../lib/ERC20SafeTransfer.sol";
+import "../../lib/MathUint.sol";
+import "../../lib/MathUint96.sol";
+import "../../lib/SignatureUtil.sol";
+import "../../thirdparty/SafeCast.sol";
+import "./AmmData.sol";
+import "./AmmStatus.sol";
+import "./AmmUtil.sol";
+
+
+/// @title AmmExitRequest
+library AmmExitRequest
+{
+    using AmmStatus         for AmmData.State;
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+    using MathUint96        for uint96;
+    using SafeCast          for uint;
+    using SignatureUtil     for bytes32;
+
+    bytes32 constant public WITHDRAW_TYPEHASH = keccak256(
+        "Withdraw(address owner,uint256[] amounts,uint256 validUntil,uint256 nonce)"
+    );
+
+    bytes32 constant public POOLEXIT_TYPEHASH = keccak256(
+        "PoolExit(address owner,bool toLayer2,uint256 poolAmountIn,uint256[] minAmountsOut,uint32[] storageIDs,uint256 validUntil)"
+    );
+
+    event Withdrawal(address owner, uint[] amountOuts);
+    event PoolExitRequested(AmmData.PoolExit exit);
+    event UnlockScheduled(address owner, uint timestamp);
+
+    function unlock(AmmData.State storage S)
+        internal
+        returns (uint lockedUntil)
+    {
+        require(S.lockedUntil[msg.sender] == 0, "UNLOCKED_ALREADY");
+
+        lockedUntil = block.timestamp + AmmData.MIN_TIME_TO_UNLOCK();
+        S.lockedUntil[msg.sender] = lockedUntil;
+
+        emit UnlockScheduled(msg.sender, lockedUntil);
+    }
+
+    function withdrawFromPool(
+        AmmData.State storage S,
+        uint[]       calldata amounts,
+        bytes        calldata signature, // signature from Exchange operator
+        uint                  validUntil
+        )
+        public
+        returns (uint[] memory amountOuts)
+    {
+        uint size = S.tokens.length;
+        require(amounts.length == size + 1, "INVALID_DATA");
+
+        _proxcessExchangeWithdrawalApprovedWithdrawals(S);
+
+        bool approvedByOperator = _checkOperatorApproval(
+            S, amounts, signature, validUntil
+        );
+
+        amountOuts = new uint[](size + 1);
+        amountOuts[0] = _withdrawToken(
+            S, address(this), amounts[0], approvedByOperator
+        );
+
+        for (uint i = 0; i < size; i++) {
+            amountOuts[i + 1] = _withdrawToken(
+                S, S.tokens[i].addr, amounts[i + 1], approvedByOperator
+            );
+        }
+
+        emit Withdrawal(msg.sender, amountOuts);
+    }
+
+    function exitPool(
+        AmmData.State storage S,
+        uint                  poolAmountIn,
+        uint96[]     calldata minAmountsOut,
+        bool                  toLayer2,
+        uint96                fee
```

Comment:
A non-zero fee is non-trivial because `fee <= amountWithdrawn` otherwise it's impossible so we'd have to check that.

But the main problem is that the operator is _required_ to process this onchain exit request regardless of the fee (so no reason to provide a non-zero fee). A fixed ETH fee like we do for forced withdrawals seems necessary while just setting `fee := 0`.

Reply:
I removed exit fees.